### ISSUE TITLE
Fake async RPCs behave very differently

### ISSUE THREAD

**Issue by evanj** [GoogleCloudPlatform/appengine-python-vm-runtime#67]
TL;DR: The README should have a caveat added about the fact that RPCs must call .wait due to the different async RPC behaviour. Maybe something like:

* Currently asynchronous RPCs are not implemented. They are emulated by calling the RPC synchronously from the `.wait()` method. If you do not call `.wait()` on an RPC, the RPC will not be sent. If your code is not properly calling `.wait()`, it is broken and will fail on rare occasions in the Standard Environment. In this new environment, it will fail far more often.

Longer story:

I'm testing moving some AppEngine code to the "python-compat" environment, which seems to be getting the library implementation that is contained in this repository. It turns out we had some horrifically thread-unsafe code that was dropping queued async RPCs and not calling .wait() on them. In the Standard Environment, I think these are fast enough (adding an item to a task queue) that they work by chance. In the new GCE environment, this code drops nearly all tasks. 

This is definitely a bug in our code, but it was revealed by the substantial behavior difference. I have a crazy proposal: Maybe the deferred RPC implementation should *make* the RPC when it is created, and simply cache the result/exception/callback until the user calls `get_result()` or `check_success()`. This would likely have made this code continue to work.

... although on the other hand, our code was terribly buggy and I'm glad we fixed it. However, this was a difference we didn't catch until we moved our code into production.

Anyway: Feel free to close this as "wont fix", but at least adding a note in the README would be great!

Thanks!

**Comment by jonparrott** [comment]
@bryanmau1 if you think this is just a documentation issue, please feel free to assign it back to me.

**Comment by jonparrott** [comment]
@evanj thanks for reporting this. This is a rather interesting use case. Do you mind sharing some code with us that demonstrates this issue? Which particular RPCs are you calling directly?

**Comment by evanj** [comment]
I just realize I may have read the code incorrectly. I need to copy the code off a GCE `python-compat` instance and add some logging to find out. :) Here is what we are doing at Bluecore, and what I *thought* the bug was. In the mean time, I'll try to investigate what is happening more carefully to see if this is how we were losing tons of queue RPCs. Thank for responding so quickly!

We are doing something like:

```queue = taskqueue.Queue('somequeue')
task = taskqueue.Task(...)
rpcs = queue.add_async(task)
```

then later we have code that fails to call .wait on these RPCs (due adding them to a global shared list that gets cleared in a thread-unsafe way). Maybe I'm tracing through the code incorrectly, but on my reading this is a non-async call in the new runtime. 

1. `taskqueue._MakeAsyncCall` calls `create_rpc` then `rpc.make_call`. create_rpc creates a `apiproxy_stub_map.UserRPC`:

https://github.com/GoogleCloudPlatform/appengine-python-vm-runtime/blob/28a49d69f5e9f5b562f14e346823a3b155626bae/appengine-compat/exported_appengine_sdk/google/appengine/api/taskqueue/taskqueue.py#L422

2. `apiproxy_stub_map.UserRPC` calls `CreateRPC`, which uses apiproxy to create some RPC implementation.

3. `UserRPC.make_call` calls `__rpc.MakeCall` in apiproxy_rpc.py. This calls `self._MakeCallImpl`. The version in apiproxy_rpc.py clearly states "Override this method to implement a real asynchronous call rpc". There is a version in vmstub.py that actually does implement asynchronous RPCs, so maybe we are using this version and our bug is somewhere else.

**Comment by bryanmau1** [comment]
Your last sentence is correct.  Check out this line.  

https://github.com/GoogleCloudPlatform/appengine-python-vm-runtime/blob/master/appengine-compat/exported_appengine_sdk/google/appengine/ext/vmruntime/vmservice.py#L135

Essentially, the apiproxy is replaced by the RPC in vmstub.

**Comment by evanj** [comment]
Oops, yes, I totally misunderstood this, and in testing this: this is clearly NOT our bug. I'm tracking down a regression that I think is unrelated. I'll file another issue if I manage to root cause it correctly. Sorry for the distraction!

**Comment by jonparrott** [comment]
@evanj, please do. We're very interested in any differences between this runtime and the GAE Standard runtime.
